### ISSUE TITLE
Mutation merging involving dead row marker and shadowable tombstones but no row tombstone is not associative

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#11307]
Given 3 row mutations:
```    
m1 = {
      marker: {row_marker: dead timestamp=-3},
      tombstone: {row_tombstone: {shadowable tombstone: timestamp=-7, deletion_time=0}, {tombstone: none}}
}

m2 = {
      marker: {row_marker: timestamp=-5}
}

m3 = {
      tombstone: {row_tombstone: {shadowable tombstone: timestamp=-6, deletion_time=2}, {tombstone: none}}
}
```

We get different shadowable tombstones depending on the order of merging:
```
(m1 + m2) + m3 = {
       marker: {row_marker: dead timestamp=-3},
       tombstone: {row_tombstone: {shadowable tombstone: timestamp=-6, deletion_time=2}, {tombstone: none}}

m1 + (m2 + m3) = {
       marker: {row_marker: dead timestamp=-3},
       tombstone: {row_tombstone: {shadowable tombstone: timestamp=-7, deletion_time=0}, {tombstone: none}}
}
```

The reason is that in the second case the shadowable tombstone in m3 is shadwed by the row marker in m2. In the first case, the marker in m2 is cancelled by the dead marker in m1, so shadowable tombstone in m3 is not cancelled (the marker in m1 does not cancel because it's dead).

This wouldn't happen if the dead marker in m1 was accompanied by a hard tombstone of the same timestamp, which would effectively make the difference in shadowable tombstones irrelevant.

Found by row_cache_test.cc::test_concurrent_reads_and_eviction.

I'm not sure if this situation can be reached in practice (dead marker in mv table but no row tombstone). \cc @nyh @psarna

**Comment by tgrabiec** [comment]
One case I'm not sure about is when the marker is expiring. Eventually it will become dead, and we will be subject to this scenario.

**Comment by tgrabiec** [comment]
Mutation compaction will turn such a marker into a virtual row tombstone by the means of this code:

```
static bool dead_marker_shadows_row(const schema& s, column_kind kind, const row_marker& marker) {
    return s.is_view()
            && s.view_info()->has_base_non_pk_columns_in_view_pk()
            && !marker.is_live()
            && kind == column_kind::regular_column; // not applicable to static rows
}
```

Does that cover all the cases?

**Comment by psarna** [comment]
Following a private conversation that ultimately led to Tomek finding dead_marker_shadows_row (https://github.com/scylladb/scylladb/issues/11307#issuecomment-1218242673), it was definitely meant to cover all the cases by the code author[1], and it does look like `compact_and_expire` is always performed on a mutation before it's used anywhere. So I'm moderately sure all cases are covered in user-facing code and we're safe.


[1] https://github.com/scylladb/scylladb/blob/516089beb0b8960ee5a9b9b73e8664e37e2275ac/mutation_partition.cc#L1666
```
commit 67dac67c46e9875979f7ce07e16af2b91a438569
Author: Duarte Nunes <duarte@scylladb.com>
Date:   Mon Apr 16 00:32:20 2018 +0100

    mutation_partition: Regular base column in view determines row liveness
    
    When views contain a primary key column that is not part of the base
    table primary key, that column determines whether the row is live or
    not. We need to ensure that when that cell is dead, and thus the
    derived row marker, either by normal deletion of by TTL, so is the
    rest of the row.
    
    This patch introduces the idea of shawdowing row marker. We map the
    status of the regular base column in the view's PK to the view row's
    marker. If this marker is dead, so is that cell in the base table, and
    so should the view row become. To enforce that, a view row's dead
    marker shadows the whole row if that view includes a base regular
    column in its PK.
    
    Fixes #3360
    
    Signed-off-by: Duarte Nunes <duarte@scylladb.com>
```
67dac67c46e9875979f7ce07e16af2b91a438569

**Comment by psarna** [comment]
Let's close this issue, it can be revised if somebody finds an issue with shadowable tombstones. Code inspection suggests that all existing cases are covered
